### ISSUE TITLE
To much synchronized result in deadlocks

### ISSUE THREAD

**Issue by nenick** [robolectric/robolectric#1043]
So where should i start? please fix my deadlock or i will do it by self ;)

I found two situations for deadlocks, and both situations first was rarely. But with speeding up the test (not more sleeping until all background should be finished) they occur more and more times until they reached a 90-100% chance.

Now both are "good" reproducible, so i could analyse them. I found some reasons/solutions and like to share them here. And like to read about your opinion to the situations and solutions.

The first deadlock was two database operations: 

The first database operation comes from the UiThread and the second operation one from Background Executor. All is fine, until one of them is not finished when the second comes. The first operation go through the UiThread Scheduler (lock) and at end it waits for a SQLiteConnection (wait). The second one Has already the SQLiteConnection (lock) und will now access the UiThread Scheduler, to get the current time (wait).

Fast solution: 
I provided my on Scheduler implementation, where the current time value is synchronized by a AtomicLong and the synchronized removed from the getCurrentTime() method. Then no more database deadlock occur.

The second deadlock is the same situation like in this issue #535

I think the same approach will work here too. Making the getShadowConfig() not synchronized with the class, instead make only the method synchronized with the "synchronized(object) {}" block. Not tried yet but i will do it in the next days.

So what do you think about the synchronisation and the reduction of them? 

I can prepare this changes and make a pull request, but it may only be unit tested with a full constructed situation. The real situation need sometimes special environment to have the 100% chance.

Just for reporting, the database deadlock thread dump

```
Found one Java-level deadlock:

"pool-3-thread-7":
  waiting to lock monitor 0x00007f3e34006390 (object 0x00000000fafda8d0, a org.robolectric.util.Scheduler),
  which is held by "Test worker"
"Test worker":
  waiting to lock monitor 0x00007f3e34005280 (object 0x00000000fa7a0770, a java.lang.Object),
  which is held by "pool-3-thread-7"

Java stack information for the threads listed above:

"pool-3-thread-7":
    at org.robolectric.util.Scheduler.getCurrentTime(Scheduler.java:16)
    - waiting to lock <0x00000000fafda8d0> (a org.robolectric.util.Scheduler)
    at org.robolectric.shadows.ShadowSystemClock.now(ShadowSystemClock.java:15)
    at org.robolectric.shadows.ShadowSystemClock.uptimeMillis(ShadowSystemClock.java:30)
    at sun.reflect.GeneratedMethodAccessor69.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.robolectric.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:473)
    at android.os.SystemClock.uptimeMillis(SystemClock.java)
    at android.database.sqlite.SQLiteConnectionPool.$$robo$$SQLiteConnectionPool_6a34_waitForConnection(SQLiteConnectionPool.java:621)
    - locked <0x00000000fa7a0770> (a java.lang.Object)
    at android.database.sqlite.SQLiteConnectionPool.waitForConnection(SQLiteConnectionPool.java)
    at android.database.sqlite.SQLiteConnectionPool.$$robo$$SQLiteConnectionPool_6a34_acquireConnection(SQLiteConnectionPool.java:348)
    at android.database.sqlite.SQLiteConnectionPool.acquireConnection(SQLiteConnectionPool.java)
    at android.database.sqlite.SQLiteSession.$$robo$$SQLiteSession_4dbc_acquireConnection(SQLiteSession.java:894)
    at android.database.sqlite.SQLiteSession.acquireConnection(SQLiteSession.java)
    at android.database.sqlite.SQLiteSession.$$robo$$SQLiteSession_4dbc_prepare(SQLiteSession.java:586)
    at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java)
    at android.database.sqlite.SQLiteProgram.$$robo$$SQLiteProgram_e9aa___constructor__(SQLiteProgram.java:58)
    at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:41)
    at android.database.sqlite.SQLiteStatement.<init>(SQLiteStatement.java:31)
    at android.database.sqlite.SQLiteDatabase.$$robo$$SQLiteDatabase_ab15_delete(SQLiteDatabase.java:1494)
    at android.database.sqlite.SQLiteDatabase.delete(SQLiteDatabase.java)
    at MyDatabaseOperation
    at org.robolectric.shadows.ShadowContentResolver.delete(ShadowContentResolver.java:247)
    at sun.reflect.GeneratedMethodAccessor98.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.robolectric.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:473)
    at android.content.ContentResolver.delete(ContentResolver.java)
    at MyDatabaseOperation
    at org.androidannotations.api.BackgroundExecutor$Task.run(BackgroundExecutor.java:302)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:334)
    at java.util.concurrent.FutureTask.run(FutureTask.java:166)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
    at java.lang.Thread.run(Thread.java:722)
"Test worker":
    at android.database.sqlite.SQLiteConnectionPool.$$robo$$SQLiteConnectionPool_6a34_releaseConnection(SQLiteConnectionPool.java:365)
    - waiting to lock <0x00000000fa7a0770> (a java.lang.Object)
    at android.database.sqlite.SQLiteConnectionPool.releaseConnection(SQLiteConnectionPool.java)
    at android.database.sqlite.SQLiteSession.$$robo$$SQLiteSession_4dbc_releaseConnection(SQLiteSession.java:906)
    at android.database.sqlite.SQLiteSession.releaseConnection(SQLiteSession.java)
    at android.database.sqlite.SQLiteSession.$$robo$$SQLiteSession_4dbc_prepare(SQLiteSession.java:590)
    at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java)
    at android.database.sqlite.SQLiteProgram.$$robo$$SQLiteProgram_e9aa___constructor__(SQLiteProgram.java:58)
    at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:41)
    at android.database.sqlite.SQLiteQuery.<init>(SQLiteQuery.java:37)
    at android.database.sqlite.SQLiteDirectCursorDriver.$$robo$$SQLiteDirectCursorDriver_7ac1_query(SQLiteDirectCursorDriver.java:44)
    at sun.reflect.GeneratedMethodAccessor110.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.fest.reflect.method.Invoker.invoke(Invoker.java:112)
    at MyDatabaseOperation
    at sun.reflect.GeneratedMethodAccessor109.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.robolectric.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:473)
    at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java)
    at android.database.sqlite.SQLiteDatabase.$$robo$$SQLiteDatabase_ab15_rawQueryWithFactory(SQLiteDatabase.java:1314)
    at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java)
    at android.database.sqlite.SQLiteDatabase.$$robo$$SQLiteDatabase_ab15_queryWithFactory(SQLiteDatabase.java:1161)
    at android.database.sqlite.SQLiteDatabase.queryWithFactory(SQLiteDatabase.java)
    at android.database.sqlite.SQLiteDatabase.$$robo$$SQLiteDatabase_ab15_query(SQLiteDatabase.java:1032)
    at android.database.sqlite.SQLiteDatabase.query(SQLiteDatabase.java)
    at android.database.sqlite.SQLiteDatabase.$$robo$$SQLiteDatabase_ab15_query(SQLiteDatabase.java:1200)
    at android.database.sqlite.SQLiteDatabase.query(SQLiteDatabase.java)
    at MyDatabaseOperation
    at org.robolectric.shadows.ShadowContentResolver.query(ShadowContentResolver.java:145)
    at sun.reflect.GeneratedMethodAccessor108.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.robolectric.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:473)
    at android.content.ContentResolver.query(ContentResolver.java)
    at android.support.v4.content.CursorLoader.$$robo$$CursorLoader_3b50_loadInBackground(CursorLoader.java:49)
    at android.support.v4.content.CursorLoader.loadInBackground(CursorLoader.java)
    at android.support.v4.content.CursorLoader.loadInBackground(CursorLoader.java:35)
    at org.robolectric.shadows.ShadowAsyncTaskLoader$BackgroundWorker.call(ShadowAsyncTaskLoader.java:49)
    at org.robolectric.util.SimpleFuture.run(SimpleFuture.java:52)
    - locked <0x00000000f2e25668> (a org.robolectric.shadows.ShadowAsyncTaskLoader$1)
    at org.robolectric.shadows.ShadowAsyncTaskLoader$2.run(ShadowAsyncTaskLoader.java:42)
    at org.robolectric.util.Scheduler.postDelayed(Scheduler.java:37)
    - locked <0x00000000fafe7970> (a org.robolectric.util.Scheduler)
    at org.robolectric.util.Scheduler.post(Scheduler.java:42)
    - locked <0x00000000fafe7970> (a org.robolectric.util.Scheduler)
    at org.robolectric.shadows.ShadowAsyncTaskLoader.onForceLoad(ShadowAsyncTaskLoader.java:40)
    at sun.reflect.GeneratedMethodAccessor200.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.robolectric.bytecode.ShadowWrangler$ShadowMethodPlan.run(ShadowWrangler.java:473)
    at android.support.v4.content.AsyncTaskLoader.onForceLoad(AsyncTaskLoader.java)
    at android.support.v4.content.Loader.$$robo$$Loader_159a_forceLoad(Loader.java:218)
    at android.support.v4.content.Loader.forceLoad(Loader.java)
    at android.support.v4.content.CursorLoader.$$robo$$CursorLoader_3b50_onStartLoading(CursorLoader.java:121)
    at android.support.v4.content.CursorLoader.onStartLoading(CursorLoader.java)
    at android.support.v4.content.Loader.$$robo$$Loader_159a_startLoading(Loader.java:198)
    at android.support.v4.content.Loader.startLoading(Loader.java)
    at android.support.v4.app.LoaderManagerImpl$LoaderInfo.$$robo$$LoaderInfo_6faa_start(LoaderManager.java:265)
    at android.support.v4.app.LoaderManagerImpl$LoaderInfo.start(LoaderManager.java)
    at android.support.v4.app.LoaderManagerImpl.$$robo$$LoaderManagerImpl_c81b_installLoader(LoaderManager.java:513)
    at android.support.v4.app.LoaderManagerImpl.installLoader(LoaderManager.java)
    at android.support.v4.app.LoaderManagerImpl.$$robo$$LoaderManagerImpl_c81b_createAndInstallLoader(LoaderManager.java:500)
    at android.support.v4.app.LoaderManagerImpl.createAndInstallLoader(LoaderManager.java)
    at android.support.v4.app.LoaderManagerImpl.$$robo$$LoaderManagerImpl_c81b_initLoader(LoaderManager.java:553)
    at android.support.v4.app.LoaderManagerImpl.initLoader(LoaderManager.java)
    at MyDatabaseOperation
    at org.robolectric.util.Scheduler$PostedRunnable.run(Scheduler.java:162)
    at org.robolectric.util.Scheduler.runOneTask(Scheduler.java:107)
    - locked <0x00000000fafda8d0> (a org.robolectric.util.Scheduler)
    at org.robolectric.util.Scheduler.advanceTo(Scheduler.java:92)
    - locked <0x00000000fafda8d0> (a org.robolectric.util.Scheduler)
    at org.robolectric.util.Scheduler.advanceToLastPostedRunnable(Scheduler.java:68)
    - locked <0x00000000fafda8d0> (a org.robolectric.util.Scheduler)
    at org.robolectric.util.Scheduler.unPause(Scheduler.java:25)
    - locked <0x00000000fafda8d0> (a org.robolectric.util.Scheduler)
    at org.robolectric.shadows.ShadowLooper.unPause(ShadowLooper.java:228)
    at org.robolectric.shadows.ShadowLooper.runPaused(ShadowLooper.java:267)
    at org.robolectric.util.ActivityController.invokeWhilePaused(ActivityController.java:229)
    at org.robolectric.util.ActivityController.start(ActivityController.java:168)
    at MyDatabaseOperation
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
    at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
    at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
    at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
    at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
    at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
    at org.robolectric.RobolectricTestRunner$2.evaluate(RobolectricTestRunner.java:250)
    at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
    at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
    at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
    at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
    at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
    at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
    at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
    at org.robolectric.RobolectricTestRunner$1.evaluate(RobolectricTestRunner.java:177)
    at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:86)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:49)
    at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:69)
    at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:50)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    at org.gradle.messaging.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
    at org.gradle.messaging.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
    at $Proxy2.processTestClass(Unknown Source)
    at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:103)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:601)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
    at org.gradle.messaging.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
    at org.gradle.messaging.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:355)
    at org.gradle.internal.concurrent.DefaultExecutorFactory$StoppableExecutorImpl$1.run(DefaultExecutorFactory.java:64)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1110)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:603)
    at java.lang.Thread.run(Thread.java:722)
´´´

```

**Comment by perlow** [comment]
"Robolectric as it is really only concerned with supporting single threaded tests."

I don't think this statement is consistent with code that purports to be thread safe and uses locks. Either you're not thread safe and you don't use locks or you're thread safe, in which case, a deadlock is a bug.

You don't call arbitrary code while holding the lock. That's right out of the textbook of how to cause deadlocks.

**Comment by nenick** [comment]
For me statement was correct. You can have only stable tests if you don't have async code execution or you are forced to adopt a concept like IdlingResources from espresso framework. Robobolectric is mainly designed for synchron test execution and you should try to archive this ;)

My problem was the library androidannotations which don't use async task for background execution. Instead it spawns a thread directly https://github.com/excilys/androidannotations/blob/develop/AndroidAnnotations/androidannotations-api/src/main/java/org/androidannotations/api/BackgroundExecutor.java which caused async code execution. 

My solution is to shadow this BackgroundExecutor and execute the async code with robolectric `ShadowApplication.getInstance().getBackgroundScheduler().postDelayed(task, 0);` and since this workaround i never got this deadlock again.

**Comment by perlow** [comment]
We reproduce the deadlock simply by calling SystemClock.elapsedRealtime on multiple threads running during the test.

**Comment by nenick** [comment]
Perhaps you can also avoid the "multiple threads"? Which libraries do you use? Show us your thread dump with the deadlock.

**Comment by jongerrish** [comment]
Hopefully fixed by #2982, please reopen if not.

**Comment by kevinvandenbreemen** [comment]
Hey there,
A team I'm on encountered this issue.  It looks like under certain circumstances the ShadowMessageQueue.enqueueMessage() can indirectly deadlock itself against MessageQueue.remove() method because the latter contains a 
synchronized(this) block...
